src/main/java/com/feedzai/commons/sql/abstraction/engine/AbstractTranslator.java

```diff
@@ -492,6 +493,54 @@ public String translate(final Values.Row row) {
         return row.isEnclosed() ? "(" + translation + ")" : translation;
     }
 
+    /**
+     * Transform values' rows into a union.
+     *
+     * @param rows the values' rows
+     * @return the resulting union
+     */
+    protected Union rowsToUnion(final List<Expression> rows) {
+        return toUnionTree(rows);
+    }
+
+    /**
+     * Create an union in form of a binary tree from a list of rows.
+     * The tree shape is to prevent stack overflow on some
+     * database engines when the list of rows is too big.
+     *
+     * @param rows the list of rows.
+     * @return an union in form of a binary tree.
+     */
+    private Union toUnionTree(final List<Expression> rows) {
+
+        List<Expression> rowsWithSelect = new ArrayList<>(rows);
+        
+        while (rowsWithSelect.size() > 2) {
```

Comment:
Interesting. My first approach would be to use recursion.  
But AFAIK this implementation is is also O(n log n), let me check.

Reply:
I usually only write something recursive when it's supernatural. Otherwise, I stick with an iterative approach.